man/sandboxfs.1
.\" Copyright 2017 Google Inc.
.\"
.\" Licensed under the Apache License, Version 2.0 (the "License"); you may not
.\" use this file except in compliance with the License.  You may obtain a copy
.\" of the License at:
.\"
.\"     http://www.apache.org/licenses/LICENSE-2.0
.\"
.\" Unless required by applicable law or agreed to in writing, software
.\" distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
.\" WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
.\" License for the specific language governing permissions and limitations
.\" under the License.
.Dd February 26, 2020
.Dt SANDBOXFS 1
.Os
.Sh NAME
.Nm sandboxfs
.Nd A virtual file system for sandboxing
.Sh SYNOPSIS
.Nm
.Op Fl -allow Ar who
.Op Fl -cpu_profile Ar path
.Op Fl -input Ar path
.Op Fl -help
.Op Fl -mapping Ar type:mapping:target
.Op Fl -node_cache
.Op Fl -output Ar path
.Op Fl -reconfig_threads Ar count
.Op Fl -ttl Ar duration
.Op Fl -version
.Op Fl -xattrs
.Ar mount_point
.Sh DESCRIPTION
.Nm
is a FUSE file system that exposes a combination of multiple files and
directories from the host's file system in the form of a virtual tree with an
arbitrary layout.
You can think of a sandbox as an arbitrary view into the host's file system with
different access privileges per directory.
.Pp
.Nm
is designed to allow running commands with limited access to the file system by
using the virtual tree as their new root, and to do so consistently across a
variety of platforms.
.Pp
The sandbox instance mounted at
.Ar mount_point
is initially configured with the mapping specifications provided via repeated
instances of
.Fl -mapping .
Once running, the mount point can be reconfigured any number of times:
.Nm
listens for reconfiguration requests on the file specified by
.Fl -input
and reconfigures the mappings as requested, emiting a response on the file
specified by
.Fl -output .
.Pp
The following flags are recognized:
.Bl -tag -width XXXX
.It Fl -allow Ar who
Specifies who should have access to the file system.
.Ar who
must be one of:
.Sq other
to indicate that everyone, including root, can access the file system;
.Sq root
to indicate that only the current user and root can access the file system; and
.Sq self
to indicate that only the current user can access the file system.
.Pp
The default value is
.Sq self
because the standard FUSE configuration does not allow more relaxed
permissions, and it would be a pity if you couldn't run
.Nm
successfully without additional configuration.
To change this behavior, add
.Sq user_allow_other
to
.Pa /etc/fuse.conf
on Linux or set the
.Sq vfs.generic.osxfuse.tunables.allow_other
.Xr sysctl 8
tunable to 1 on macOS.
.Pp
Note that, at least on macOS, you will need to consider granting
.Sq other
permissions if you intend to run signed binaries through the sandbox.
This is because the
.Xr amfid 8
daemon, which implements the signature validation, runs as a different user
and must be able to access the executables.
.It Fl -cpu_profile Ar path
Enables CPU profiling and stores the pprof log to the given
.Ar path .
This feature is only available if
.Nm
was built with gperftools support (i.e. with the compile-time
.Sq profiler
feature).
Passing this flag when support is not enabled results in an error.
.It Fl -input Ar path
Points to the file from which to read new configuration requests, or
.Sq -
(the default) for stdin.
See the
.Sx Reconfigurations
subsection for details on the contents and behavior of the input file.
.It Fl -help
Prints global help details and exits.
Specifying this flag causes all other valid flags and arguments to be ignored.
.It Fl -mapping Ar type:mapping:target
Registers a new mapping.
This flag can be given an arbitrary number of times as long as the same
.Ar mapping
is not repeated.
See the
.Sx Mapping specifications
subsection for details on how a mapping is specified.
.It Fl -node_cache
Enables the path-based node cache, which causes nodes to be reused across
reconfigurations when they map to the same underlying paths.
This should offer a performance boost when the same set of files are mapped
over and over again across different reconfiguration operations.
.Pp
This cache used to be enabled by default but it turned out to be problematic.
For example, on macOS, OSXFUSE seems to keep track of the path where a node
first appeared, which causes trouble if the node disappears and then reappears
in a different location (for example,
.Xr dladdr 3
returns invalid paths).
For this reason, the cache is now disabled by default but is still kept around
in the code for experimentation.
.Em The cache may be removed in a future release altogether if these problems
.Em cannot be worked around.
.It Fl -output Ar path
Points to the file to which to write confirmations of reconfiguration, or
.Sq -
(the default) for stdout.
See the
.Sx Reconfigurations
subsection for details on the contents and behavior of the output file.
.It Fl -ttl Ar duration
Specifies how long the kernel is allowed to cache file metadata for.
The duration is currently specified as a number of seconds followed by the
.Sq s
suffix.
.It Fl -reconfig_threads Ar count
Sets the number of threads to use to process reconfiguration requests.
Defaults to the number of logical CPUs in the system.
.It Fl -version
Prints version information and exits.
Specifying this flag causes all other valid flags and arguments to be ignored
except for
.Fl -help ,
which has priority over this one.
.Pp
The first line printed to stdout is machine parseable and follows the following
structure:
.Bd -literal -offset indent
sandboxfs MAJOR.MINOR[.ADDITIONAL OPTIONAL QUALIFIERS]
.Ed
.Pp
Programs automating invocations of
.Nm
can use this information to determine the correct command-line syntax to use.
.It Fl -xattrs
Enables support for extended attributes, which causes all extended attribute
operations to propagate to the underlying files.
.Pp
When disabled (the default), the behavior of extended attribute operations is
platform-dependent: they may either fail as not supported or they may act as
no-op stubs.
The reason is that this feature is disabled by telling the kernel that none of
these operations are implemented in
.Nm ,
which then causes the kernel to never contact the daemon for them.
Keeping extended attributes disabled can result in a performance boost.
.El
.Ss Mapping specifications
The basic concept behind a
.Nm
instance is the
.Em mapping specification .
A mapping specification defines how an arbitrary location within the sandbox
relates to another arbitrary location on the host file system.
Mapping specifications have three components: the
.Em target ,
which is a path to a file or directory into the host file system (i.e. outside
of the mount point); the
.Em mapping ,
which is an absolute path within the mount point in which the target is exposed;
and the
.Em type ,
which specifies the permissions of the mapping.
.Pp
The target must exist but the mapping may not: in particular, any path
subcomponents of the mapping that have not been previously mapped
within the sandbox will be created as virtual read-only nodes.
This means that the ordering of the mappings matters.
.Pp
For example: consider a
.Nm
instance mounted onto
.Pa /mnt
that maps
.Pa /my/custom/bin
to the
.Pa /usr/bin
target.
With this setup,
.Pa /mnt/my/custom/bin
will expose the contents of the real
.Pa /usr/bin
directory.
.Pp
The following types are currently supported:
.Bl -tag -width XXXX
.It ro
A read-only mapping.
The contents of the target are exposed verbatim at the mapping point and they
cannot be modified through the mount point.
Any write access will result in an
.Dv EPERM .
.It rw
A read/write mapping.
The contents of the target are exposed verbatim at the mapping point and they
can be modified at will through the mount point.
Writes through the moint point are applied immediately to the underlying target
directory.
.El
.Ss Reconfigurations
While a mount point is live,
.Nm
listens for reconfiguration requests on the file specified by
.Fl -input
and writes reconfiguration responses to the file specified by
.Fl -output .
These two files essentially implement a rudimentary RPC system subject to
change.
.Pp
The mount point can be configured any number of times while mounted,
which allows for processes to efficiently change the view of the sandbox at will
without having to remount it.
Closing the input file causes
.Nm
to freeze its configuration and not accept any more requests for
reconfiguration, but the file system will continue to operate normally until
it is either unmounted or signaled.
.Pp
Reconfiguration requests can be issued at any time.
However, it is impossible for
.Nm
to properly deal with ongoing file system operations or with open file handles
in a deterministic manner.
Due to this, the behavior of ongoing file system activity while a
reconfiguration happens is unspecified but can be assumed to be stable (i.e.
.Nm
will not deadlock nor crash).
As a result, it is highly recommended that you only send reconfiguration
requests when you know that the file system is quiescent.
.Pp
Configuration requests are provided as a stream of JSON objects.
Each request is an object with just one of the following keys:
.Sq CreateSandbox ,
which requests the creation of a new top-level directory with a given set of
mappings; and
.Sq DestroySandbox ,
which requests the deletion of the mappings at an existing top-level directory.
.Pp
.Pp
A
.Sq CreateSandbox
operation contains an object with three keys:
.Sq id
specifies the name of the top-level directory to create,
.Sq mappings
contains an array of all mappings to apply within that directory, and
.Sq prefixes
contains a dictionary of all new prefixes to register for prefix-encoded paths.
Each mapping entry is an object with the following keys:
.Sq path
and
.Sq underlying_path ,
which define the mapping paths;
.Sq path_prefix
and
.Sq underlying_path_prefix ,
which identify the prefixes for the provided paths, respectively; and
.Sq writable ,
which if set to true indicates a read/write mapping.
The mapping must not yet exist in the file system.
Each entry in the prefixes dictionary is keyed by the numerical identifier of
the prefix (supplied as a string due to JSON limitations), and the value is the
absolute path for that prefix.
.Pp
Prefix-encoded paths serve to minimize the size of the requests but their use
is optional.
To specify a path in its prefix-encoded form,
.Sq path
and/or
.Sq underlying_path
must be relative and
.Sq path_prefix
and/or
.Sq underlying_path_prefix
must indicate a positive non-zero integer.
The referenced prefix must be registered in the same request or a previous
request, but if the prefix is repeated across requests, it must point to the
same location.
To avoid prefix-encoded form, paths must be given as absolute and they need to
reference the special prefix 0.
.Pp
A
.Sq DestroySandbox
operation simply contains the
.Sq id
of a previously-created sandbox as a string.
The whole tree hierarchy is unmapped.
.Pp
Each configuration request is paired with a response, which are also provided
as a stream of JSON objects.
Each response is a map with an optional
.sq id
field, which corresponds to the identifier given in the request, and an optional
.Sq error
field, which is empty if the request was successful and contains an error
message otherwise.
Responses with a missing identifier indicate fatal failures during the
reconfiguration (e.g. due to a syntax error) and are not recoverable.
.Pp
To minimize the size of the requests, all fields support aliases and default
values as follows:
.Pp
.Bl -tag -offset XXXXXX -width underlyingXpathXprefixXX -compact
.It Sq CreateSandbox
Alias:
.Sq C .
.It Sq DestroySandbox
Alias:
.Sq D .
.It Sq id
Alias:
.Sq i .
.It Sq mappings
Alias:
.Sq m .
Default value: empty array.
.It Sq prefixes
Alias:
.Sq q .
Default value: empty object.
.It Sq path
Alias:
.Sq p .
.It Sq path_prefix
Alias:
.Sq x .
Default value:
.Sq 0 .
.It Sq underlying_path
Alias:
.Sq u .
.It Sq underlying_path_prefix
Alias:
.Sq y .
Default value:
.Sq 0 .
.It Sq writable
Alias:
.Sq w .
Default value:
.Sq false .
.El
.Sh EXIT STATUS
.Nm
exits with 0 if the file system was both mounted and unmounted cleanly; 1 on a
controlled error condition encountered during the execution of a command; or 2
on a usage error.
.Pp
Sending a termination signal to
.Nm
will cause the file system to exit in a controlled manner, ensuring that the
mount point is released.
If the file system is busy, the signal will be queued until all open file
descriptors on the file system are released at which point the file system
will try to exit cleanly again.
Note that, due to limitations in signal handling in Rust (which is the language
in which
.Nm
is implemented), the reception of a signal will cause
.Nm
to return 1 instead of terminating with a signal condition.
.Sh ENVIRONMENT
.Nm
recognizes the following environment variables:
.Bl -tag -width XXXX
.It Va RUST_LOG
Sets the maximum level of logging messages sent to stderr.
Possible values include
.Sq error ,
.Sq warn ,
.Sq info
and
.Sq debug ,
though many more syntaxes are supported.
See the documentation for Rust's
.Sq env_logger
crate for more details.
.El
.Pp
.Nm
may recognize other standard Rust variables like
.Va RUST_BACKTRACE
but the list above attempts to describe the ones a user may find most useful.
.Sh EXAMPLES
.Ss Command-line invocation
This example configures a sandbox that maps the whole host's
file system but clears
.Pa /tmp
to point into a sandbox-specific writable directory:
.Bd -literal -offset indent
sandboxfs --mapping=ro:/:/ --mapping=rw:/tmp:/tmp/fresh-tmp /mnt
.Ed
.Ss Reconfiguration request
This example creates a new sandbox under
.Pa /first
with a single mapping that will appear as
.Pa /first/tmp
and that is mapped to
.Pa /tmp/abc ;
then deletes said sandbox, and then creates a different sandbox under
.Pa /second
with a single mapping that will appear as
.Pa /foo/bar
and that is mapped to
.Pa /tmp/x/y :
.Bd -literal -offset indent
[
    {"CreateSandbox": {
        "id": "first",
        "mappings": [
            {"path": "/tmp", "underlying_path": "/tmp/abc", "writable": true}
        ]
    }},
    {"DestroySandbox": "first"}
    {"CreateSandbox": {
        "id": "second",
        "mappings": [
            {"path": "/foo/bar", "underlying_path": "/tmp/x/y",
             "writable": false}
        ]
    }}
]
.Ed
.Pp
When sending this request, we obtain the following response:
.Bd -literal -offset indent
{"id": "first", "error": null}
{"id": "first", "error": null}
{"id": "second", "error": null}
.Ed
.Ss Prefix-encoded reconfiguration request
This example is the same as above, but this time using prefix-encoded paths:
.Bd -literal -offset indent
[
    {"CreateSandbox": {
        "id": "first",
        "mappings": [
            {"path": "", "path_prefix": 1,
             "underlying_path": "abc", "underlying_path_prefix": 1,
             "writable": true}
        ],
        "prefixes": {
            "1": "/tmp"
        }
    }},
    {"DestroySandbox": "first"}
    {"CreateSandbox": {
        "id": "second",
        "mappings": [
            {"path": "bar", "path_prefix": 2,
             "underlying_path": "x/y", "underlying_path_prefix": 1,
             "writable": false}
        ],
        "prefixes": {
            "2": "/foo"
        }
    }}
]
.Ed
.Ss Minimized reconfiguraton request
This example is the same as above, but this time using field aliases and
default values to minimize the message:
.Bd -literal -offset indent
[
    {"C": {
        "i": "first",
        "m": [
            {"p": "", "x": 1, "u": "abc", "y": 1, "w": true}
        ],
        "q": {
            "1": "/tmp"
        }
    }},
    {"D": "first"}
    {"C": {
        "i": "second",
        "m": [
            {"p": "bar", "x": 2, "u": "x/y", "y": 1}
        ],
        "q": {
            "2": "/foo"
        }
    }}
]
.Ed
.Sh AUTHORS
.An -nosplit
The
.Nm
was originally developed as a Go program by
.An Pallav Agarwal
.Aq pallavag@google.com
with guidance from
.An Julio Merino
.Aq jmmv@google.com .
The program was later reimplemented in Rust by the latter.
.Sh BUGS
The following are known limitations of
.Nm :
.Bl -bullet
.It
Hard links are not supported.
.It
Mapping the same external file or directory under two different locations within
the mount point results in undefined behavior.
Writes may not be reflected at both mapped locations at the same time, which
can lead to data corruption.
This is true even for read-only mappings because each separate view within
the mount point may have cached different contents, returning different data
than what's truly on disk.
Using
.Fl -node_cache
may help mitigate this issue but it doesn't always do.
.It
The
.Fl -allow Ar root
setting does not work on Linux; use
.Fl -allow Ar other
as the alternative.
See https://github.com/bazil/fuse/issues/144 for details.
.It
It is currently impossible to terminate
.Nm
cleanly while the file system is busy.
Signals received while the file system is in use will be queued as described in
.Sx EXIT STATUS
and fatal signals will cause
.Nm
to leak the mount point (possibly irrecoverably without a reboot because of
kernel bugs).
.It
Any explicitly-mapped directories and any scaffold directories (those
directories that appear to represent intermediate path components that do not
exist anywhere else in the file system) cannot be removed.
Attempts to remove them will result in a
.Dq permission denied
error.
While it could be possible to implement some different behavior, this is what
.Nm
currently exposes.
You may or may not consider this to be a bug.
.It
Node data is cached in-memory for all files accessed through a
.Nm
instance in order to offer good performance across reconfigurations.
However, this cache does not currently implement any expiration policy, which
means that memory usage can grow unboundedly if many different files are
mapped and accessed through the sandbox.
.It
If a FIFO is used for
.Fl input ,
.Nm
will block until a separate process opens the FIFO for writing.
This happens even before the file system starts serving, which means the file
system is not usable until the FIFO is opened.
.It
While it is possible to reconfigure the entries of the root directory of a
running file system, it is not possible to reconfigure the root mapping itself
to point to a different location or to change its writability properties.
.It
Unmapping entries doesn't fully work.
The FUSE library that
.Nm
currently uses does not support sending cache invalidation requests to the
kernel, which means unmapped entries will not vanish immediately from the
file system.
You may be able to mitigate this by setting a low node TTL with the
.Fl -ttl
flag, but this doesn't work on macOS either because OSXFUSE does not honor
node TTLs.
.It
Handling of extended attributes on open-but-deleted-files does not work
properly.
Those files will appear as if they didn't have any extended attributes any
longer, and attempts to modify them will fail.
.El


NEWS.md
# Major changes between releases

## Changes in version 0.2.0

**STILL UNDER DEVELOPMENT; NOT RELEASED YET.**

*   Changed the reconfiguration protocol to use JSON streams for both the
    requests and the responses, instead of the previous ad-hoc line-oriented
    protocol.

*   Changed the reconfiguration protocol so that each map and unmap request
    carries a list of mappings to map and unmap, respectively, along with the
    "root" path where all those mappings start.  This is to allow sandboxfs to
    process the requests more efficiently.

*   Changed the reconfiguration protocol so that each request contains a tag,
    which is then propagated to the response for that request.  This is to
    allow sandboxfs to process requests in parallel.

*   Changed the reconfiguration protocol to work at the level of sandboxes,
    not paths, where a sandbox is defined as a top-level directory with a
    collection of mappings beneath it.

    This essentially makes reconfigurations less powerful than they were, but
    also makes them infinitely simpler to understand and manage.  Furthermore,
    this lines up better with the needs of Bazel, our primary customer, and
    with sandboxfs' own name.

*   Changed the reconfiguration protocol to take prefix-encoded paths to
    minimize the size of the reconfiguration requests.  This has shown to
    significantly reduce the CPU consumption of both sandboxfs and Bazel
    during a build, as the size of the reconfiguration messages is
    drastically smaller.

*   Changed the reconfiguration protocol to accept short aliases for all
    fields, thus further minimizing the size of reconfiguration requests,
    and also to accept omitting optional fields.

*   Made sandboxfs process reconfiguration requests in parallel, which has a
    significant performance impact when those requests are large.

*   Fixed a bug where writes on a file descriptor that had been duplicated and
    closed did not update the file size, resulting in bad data being returned
    on future reads.

*   Fixed timestamp updates so that the `birthtime` rolls back to an older
    `mtime` to mimic BSD semantics.

*   Fixed hardlink counts so that they are zero for handles that point to
    deleted files or directories.

*   Added support for extended attributes.  Must be explicitly enabled by
    passing the `--xattrs` option.

*   Added support to change the timestamps of a symlink on systems that have
    this feature.

*   Disabled the path-based node cache by default and added a `--node_cache`
    flag to reenable it.  This fixes crashes running Java within a sandboxfs
    instance where the Java toolchain is mapped under multiple locations and
    the first mapped location vanishes.  See [The OSXFUSE, hard links, and
    dladdr puzzle](https://jmmv.dev/2020/01/osxfuse-hardlinks-dladdr.html) for
    details.

## Changes in version 0.1.1

**Released 2019-10-24.**

*   Fixed the definition of `--input` and `--output` to require an argument,
    which makes `--foo bar` and `--foo=bar` equivalent.  This can be thought to
    break backwards compatibility but, in reality, it does not.  The previous
    behavior was just broken: specifying `--foo bar` would cause `bar` to be
    treated as an argument and `--foo` to use its default value, which meant
    that these two flags would be ignored when supplied under this syntax.

*   Fixed `--input` and `--output` to handle stdin and stdout correctly when
    running e.g. under `sudo`.

*   Make create operations honor the UID and GID of the caller user instead of
    inheriting the permissions of whoever was running sandboxfs.  Only has an
    effect when using `--allow=other` or `--allow=root`.

## Changes in version 0.1.0

**Released on 2019-02-05.**

This is the first formal release of the sandboxfs project.

**WARNING:** The interaction points with sandboxfs are subject to change at this
point.  In particular, the command-line interface and the data format used to
reconfigure sandboxfs while it's running *will* most certainly change.


cmd/sandboxfs/sandboxfs.go
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License.  You may obtain a copy
// of the License at:
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
// License for the specific language governing permissions and limitations
// under the License.

// The sandboxfs binary mounts an instance of the sandboxfs filesystem.
package main

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/signal"
	"path/filepath"
	"runtime"
	"syscall"
	"time"

	"bazil.org/fuse"
	"github.com/bazelbuild/sandboxfs/internal/sandbox"
)

var (
	// packageVersion contains the textual version number of this package.
	//
	// The value here is populated by the build system.  The placeholder denotes that
	// a build that misses to include this detail should not be shipped to users.
	packageVersion = "0.0 (BINARY NOT FOR RELEASE)"
)

// handleSignals installs signal handlers to ensure the file system is unmounted.
//
// The signal handler is responsible for unmounting the file system, which in turn causes the
// FUSE serve loop to either never start or to finish execution if it was already running.
//
// But this is tricky business because of a potential race: if the signal handler is installed
// and a signal arrives *before* the mount point has been configured, the unmounting will not
// succeed, which means we will enter the server loop and lose the signal. Conversely, if we
// did this backwards, we could receive a signal after the mount point has been configured but
// before we install the signal handler, which means we'd terminate but leak the mount point.
//
// To solve this, we must install the signal handler first, but we must not take action on
// signals until after we know that the mount point has been set up. This is what the mountPoint
// channel is for: the caller must inject the mount point into this handler only after the mount
// operation has succeeded, or an empty string if the mount operation failed.
//
// The caller can know if a signal was the cause of the FUSE serve loop termination by inspecting
// the contents of the caughtSignal output channel.
//
// NOTE: This would be much easier if we could just mask signals while we prepare the mount point in
// the caller and then process them once we are ready. Unfortunately, signal.Ignore() does not only
// mask signals: it discards them as well! The x/sys/unix package is supposed to gain signal
// handling at some point so maybe we'll be able to revisit this in the future? Who knows.
func handleSignals(mountPoint <-chan string, caughtSignal chan<- os.Signal) {
	handler := make(chan os.Signal, 1)
	signal.Notify(handler, syscall.SIGHUP, os.Interrupt, syscall.SIGQUIT, syscall.SIGTERM)

	go func() {
		caughtSignal <- <-handler // Wait for the signal.

		// Wait until the main process has had a chance to issue the mount(2) system call.
		// It doesn't matter if such call succeeded or not: we must wait anyway before
		// attempting to do any cleanup.
		mnt := <-mountPoint

		if mnt != "" {
			// Now the real deal: the main process did actually get to issue a
			// successful mount(2) system call. Make the mount point vanish so that the
			// FUSE serve loop terminates or prevent it from starting.
			backoff := 10 * time.Millisecond
			for {
				err := fuse.Unmount(mnt)
				if err == nil {
					break
				}

				// If unmounting fails, it is probably because the file system is
				// busy. We don't know, but it doesn't matter: we have entered a
				// terminal status: we know we have to exit, so we'll keep trying to
				// unclog things while telling the user what's going on.  They are
				// the ones that have to fix this situation.
				log.Printf("unmounting filesystem failed with error: %v; will retry in %v", err, backoff)
				time.Sleep(backoff)
				if backoff < time.Second {
					backoff = backoff * 2
				}
			}
		}
	}()
}

func serve(settings ProfileSettings, mountPoint string, options []fuse.MountOption, initialMappings []sandbox.MappingSpec, reconfigInput io.Reader, reconfigOutput io.Writer) error {
	root, err := sandbox.CreateRoot(nil, initialMappings)
	if err != nil {
		return fmt.Errorf("unable to init sandbox: %v", err)
	}

	profileContext, err := StartProfiling(settings)
	if err != nil {
		return err
	}
	defer profileContext.Close()

	// OSXFUSE unconditionally creates the mount point if it does not exist while Linux's FUSE
	// errors out on this condition. Linux is behaving correctly here, but to unify the behavior
	// between the two cases (and, especially, to ensure that the error message that we print is
	// consistent), explicitly test for the mount point's existence.
	//
	// Note that this is knowingly racy. If the mount point is created after this call but before
	// the actual mount operation happens, we'll be subject to OS-specific behavior. We cannot do
	// do better, and this is not a big deal anyway.
	//
	// TODO(jmmv): Fix OSXFUSE to not create the mount point. If that's undesirable upstream, an
	// alternative would be to add a "-o nocreate_mount" option to mount_osxfuse and then use that
	// in the fuse.Mount call below.
	if _, err := os.Lstat(mountPoint); os.IsNotExist(err) {
		return fmt.Errorf("unable to mount: %s does not exist", mountPoint)
	}

	mountOk := make(chan string, 1)
	caughtSignal := make(chan os.Signal, 1)
	handleSignals(mountOk, caughtSignal)

	c, err := fuse.Mount(mountPoint, options...)
	if err != nil {
		mountOk <- "" // Neutralize signal handler.

		// Even if fuse.Mount failed, we can still hit the case where the mount point was
		// registered with the kernel. Try to unmount it here as a best-effort operation.
		// (I.e. we can't tell upfront if the mount point was registered so we have to
		// unconditionally try to unmount it and hope it gets cleaned up.)
		//
		// This was observed to happen on Linux: mounting a FUSE file system requires
		// spawning the fusermount program aside from telling the kernel about the mount
		// point. It can happen that a signal arrives at "the wrong time" within the
		// fuse.Mount call above and we get an error here even when the mount point is left
		// behind. This is likely a bug in the fuse.Mount logic.
		fuse.Unmount(mountPoint)

		return newMountError("unable to mount: %v", err)
	}
	defer c.Close()
	mountOk <- mountPoint // Tell signal handler that the mount point requires cleanup.

	err = sandbox.Serve(c, root, reconfigInput, reconfigOutput)
	if err != nil {
		return fmt.Errorf("serve error: %v", err)
	}

	<-c.Ready
	if err := c.MountError; err != nil {
		return fmt.Errorf("mount error: %v", err)
	}

	// If we reach this point, the FUSE serve loop has terminated because the user unmounted the
	// file system or because we received a signal. In both cases we need to exit, but we treat
	// signal receipts as an error just so that the user can tell that the exit was not clean.
	select {
	case signal := <-caughtSignal:
		return fmt.Errorf("caught signal: %v", signal.String())
	default:
	}
	return nil
}

// safeMain is a version of main that does not exit on its own. Instead, it returns an error type
// so that the real main function can format all errors consistently across all commands.
func safeMain(progname string, args []string) error {
	flags := flag.NewFlagSet(progname, flag.ContinueOnError)
	flags.SetOutput(ioutil.Discard)
	flags.Usage = func() {}

	var allow allowFlag
	allow.Set("self")
	flags.Var(&allow, "allow", "specifies who should have access to the file system; must be one of other, root, or self")
	cpuProfile := flags.String("cpu_profile", "", "write a CPU profile to the given file on exit")
	debug := flags.Bool("debug", false, "log details about FUSE requests and responses to stderr")
	help := flags.Bool("help", false, "print the usage information and exit")
	input := flags.String("input", "-", "where to read the configuration data from (- for stdin)")
	listenAddress := flags.String("listen_address", "", "enable HTTP server on the given address and expose pprof data")
	var initialMappings mappingFlag
	flags.Var(&initialMappings, "mapping", "mappings of the form TYPE:MAPPING:TARGET")
	memProfile := flags.String("mem_profile", "", "write a memory profile to the given file on exit")
	output := flags.String("output", "-", "where to write the status of reconfiguration to (- for stdout)")
	version := flags.Bool("version", false, "show version information and exit")
	volumeName := flags.String("volume_name", "sandbox", "name for the sandboxfs volume")

	if err := flags.Parse(args); err != nil {
		if err == flag.ErrHelp {
			// The flags library insists on offering a -h flag even if we explicitly
			// defined --help above. Turn it into an error.
			return newUsageError("flag provided but not defined: -h")
		}
		return newUsageError("%v", err)
	}

	if *help {
		fmt.Fprintf(os.Stdout, "Usage: %s [flags...] mount-point\n\n", progname)
		fmt.Fprintf(os.Stdout, "Available flags:\n")
		flags.SetOutput(os.Stdout)
		flags.PrintDefaults()
		return nil
	}

	if *version {
		fmt.Fprintf(os.Stdout, "sandboxfs %s\n", packageVersion)
		return nil
	}

	settings, err := NewProfileSettings(*cpuProfile, *memProfile, *listenAddress)
	if err != nil {
		return newUsageError("invalid profiling settings: %v", err)
	}

	if *debug {
		fuse.Debug = func(msg interface{}) { fmt.Fprintln(os.Stderr, msg) }
	}

	if flags.NArg() != 1 {
		return newUsageError("invalid number of arguments")
	}
	mountPoint := flags.Arg(0)

	reconfigInput := os.Stdin
	if *input != "-" {
		file, err := os.Open(*input)
		if err != nil {
			return fmt.Errorf("unable to open file %q for reading: %v", *input, err)
		}
		defer file.Close()
		reconfigInput = file
	}
	reconfigOutput := os.Stdout
	if *output != "-" {
		file, err := os.Create(*output)
		if err != nil {
			return fmt.Errorf("unable to open file %q for writing: %v", *output, err)
		}
		defer file.Close()
		reconfigOutput = file
	}

	options := []fuse.MountOption{
		// Rely on in-kernel permission checking based on the node's ownership and mode to
		// avoid having to implement Access -- and we shouldn't implement it because doing
		// so causes a noticeable performance regression with OSXFUSE.
		fuse.DefaultPermissions(),

		fuse.VolumeName(*volumeName),

		// TODO(jmmv): Should be user-customizable.
		fuse.FSName("sandboxfs"),
		fuse.Subtype("sandboxfs"),

		// Do not enable fuse.LocalVolume(): doing so causes macOS to issue additional
		// operations on the file system (e.g. stats on hidden files and tree indexing),
		// which are detrimental to performance. If you feel like you have to add this
		// option, it should probably be user-customizable.
	}
	if allow.Option != nil {
		options = append(options, allow.Option)
	}

	err = serve(settings, mountPoint, options, initialMappings, reconfigInput, reconfigOutput)
	if runtime.GOOS == "linux" && allow.String() == "root" {
		if _, ok := err.(*mountError); ok {
			// "-o allow_root" is broken on Linux because this is not actually a
			// fusermount option: it is a libfuse option and the Go bindings don't
			// implement it as such.  We could implement this on our own by handling
			// allow_root as if it were allow_other with an explicit user check... but
			// it's probably not worth doing.  For now, just tell the user that we know
			// about the breakage.
			//
			// See https://github.com/bazil/fuse/issues/144 for context.
			err = newMountError("%v (-allow=root is known to be broken on Linux)", err)
		}
	}
	return err
}

func main() {
	progname := filepath.Base(os.Args[0])
	args := os.Args[1:]

	if err := safeMain(progname, args); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)

		switch err.(type) {
		case *usageError:
			fmt.Fprintf(os.Stderr, "Type '%s --help' for details\n", progname)
			os.Exit(2)
		default:
			os.Exit(1)
		}
	}
}
